lib/bolt/transport/local/shell.rb

```diff
@@ -1,26 +1,195 @@
 # frozen_string_literal: true
 
 require 'open3'
+require 'fileutils'
 require 'bolt/node/output'
+require 'bolt/util'
 
 module Bolt
   module Transport
-    class Local
-      class Shell
-        def execute(*command, options)
-          command.unshift(options[:interpreter]) if options[:interpreter]
-          command = [options[:env]] + command if options[:env]
-
-          if options[:stdin]
-            stdout, stderr, rc = Open3.capture3(*command, stdin_data: options[:stdin], chdir: options[:dir])
+    class Local < Sudoable
+      class Shell < Sudoable::Connection
+        attr_accessor :user, :logger, :target
+        attr_writer :run_as
+
+        def initialize(target)
+          @target = target
+          @user = Etc.getlogin
+          @run_as = target.options['run-as']
+          @logger = Logging.logger[self]
+        end
+
+        # If prompted for sudo password, send password to stdin and return an
+        # empty string. Otherwise, check for sudo errors and raise Bolt error.
+        # If error is not sudo-related, return the stderr string to be added to
+        # node output
+        def handle_sudo(stdin, err, pid)
+          if err.include?(Sudoable.sudo_prompt)
+            if @target.options['sudo-password']
+              # Hopefully no one's sudo-password is > 64kb
+              stdin.write("#{@target.options['sudo-password']}\n")
+              ''
+            else
+              raise Bolt::Node::EscalateError.new(
+                "Sudo password for user #{@user} was not provided for localhost",
+                'NO_PASSWORD'
+              )
+            end
+          else
+            handle_sudo_errors(err, pid)
+          end
+        end
+
+        def handle_sudo_errors(err, pid)
+          if err =~ /^#{@user} is not in the sudoers file\./
+            @logger.debug { err }
+            raise Bolt::Node::EscalateError.new(
+              "User #{@user} does not have sudo permission on localhost",
+              'SUDO_DENIED'
+            )
+          elsif err =~ /^Sorry, try again\./
+            @logger.debug { err }
+            Process.kill('TERM', pid)
```

Comment:
This doesn't work the sudo process is owned by root so it can't be killed from here instead it raises `Errno::EPERM`

Reply:
Even if the user hasn't authenticated yet?

This raises the question: I'm not sure (obviously) how to test this, or how to kill a process owned by sudo if the user can't authenticate....

Reply:
Still not addressed